File: src/core/SpectralRealityModel.js
export class SpectralObject {
  constructor({
    id,
    kind,
    origin,
    signature,
    stability,
    drift,
    confidence,
    relationships,
    metadata
  }) {
    if (!id || !kind || !origin) {
      throw new Error("SpectralObject requires id, kind, and origin.");
    }

    this.id = id;                          // Stable identifier within catalog
    this.kind = kind;                      // e.g., "dom-sheet", "json-schema", "state-machine"
    this.origin = origin;                  // e.g., { domain, system, runId, modality }
    this.signature = signature || {};      // Shape-defining fields (selectors, fields, spans, etc.)
    this.stability = typeof stability === "number" ? stability : 0.0;   // 0–1
    this.drift = typeof drift === "number" ? drift : 0.0;               // 0–1
    this.confidence = typeof confidence === "number" ? confidence : 0.0;// 0–1
    this.relationships = Array.isArray(relationships) ? relationships : [];
    this.metadata = metadata || {};        // Free-form: tags, impact, notes
    this.createdAt = new Date().toISOString();
    this.updatedAt = this.createdAt;
  }

  touch(update = {}) {
    if (typeof update.stability === "number") this.stability = update.stability;
    if (typeof update.drift === "number") this.drift = update.drift;
    if (typeof update.confidence === "number") this.confidence = update.confidence;
    if (Array.isArray(update.relationships)) this.relationships = update.relationships;
    if (update.metadata && typeof update.metadata === "object") {
      this.metadata = { ...this.metadata, ...update.metadata };
    }
    this.updatedAt = new Date().toISOString();
    return this;
  }

  toJSON() {
    return {
      id: this.id,
      kind: this.kind,
      origin: this.origin,
      signature: this.signature,
      stability: this.stability,
      drift: this.drift,
      confidence: this.confidence,
      relationships: this.relationships,
      metadata: this.metadata,
      createdAt: this.createdAt,
      updatedAt: this.updatedAt
    };
  }
}

export class SpectralRealityModel {
  constructor() {
    this.objects = new Map(); // id -> SpectralObject
  }

  upsert(raw) {
    const existing = this.objects.get(raw.id);
    if (existing) {
      existing.touch(raw);
      return existing;
    }
    const obj = new SpectralObject(raw);
    this.objects.set(obj.id, obj);
    return obj;
  }

  getById(id) {
    return this.objects.get(id) || null;
  }

  listByKind(kind) {
    const out = [];
    for (const obj of this.objects.values()) {
      if (obj.kind === kind) out.push(obj);
    }
    return out;
  }

  listHighStability(threshold = 0.8) {
    const out = [];
    for (const obj of this.objects.values()) {
      if (obj.stability >= threshold && obj.drift <= 1 - threshold) {
        out.push(obj);
      }
    }
    return out;
  }

  listByOriginDomain(domain) {
    const out = [];
    for (const obj of this.objects.values()) {
      if (obj.origin && obj.origin.domain === domain) out.push(obj);
    }
    return out;
  }

  snapshot() {
    return Array.from(this.objects.values()).map(o => o.toJSON());
  }
}

File: src/core/spectral_reality_model.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use uuid::Uuid;

/// Kinds of spectral‑objects that can be excavated.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum SpectralKind {
    DomSheet,
    JsonSchema,
    StateMachine,
    ApiShape,
    TracePattern,
    VmRegion,
    Other(String),
}

/// Origin of a spectral‑object.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Origin {
    pub domain: String,
    pub system: String,
    pub run_id: String,
    pub modality: String,
}

/// Core spectral‑object representation.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SpectralObject {
    pub id: String,
    pub kind: SpectralKind,
    pub origin: Origin,
    pub signature: HashMap<String, serde_json::Value>,
    pub stability: f64,
    pub drift: f64,
    pub confidence: f64,
    pub relationships: Vec<String>,
    pub metadata: HashMap<String, serde_json::Value>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

impl SpectralObject {
    /// Creates a new spectral‑object.
    pub fn new(
        kind: SpectralKind,
        origin: Origin,
        signature: HashMap<String, serde_json::Value>,
    ) -> Self {
        let now = Utc::now();
        SpectralObject {
            id: Uuid::new_v4().to_string(),
            kind,
            origin,
            signature,
            stability: 0.0,
            drift: 0.0,
            confidence: 0.0,
            relationships: Vec::new(),
            metadata: HashMap::new(),
            created_at: now,
            updated_at: now,
        }
    }

    /// Updates mutable fields and touches the timestamp.
    pub fn touch(
        &mut self,
        stability: Option<f64>,
        drift: Option<f64>,
        confidence: Option<f64>,
        relationships: Option<Vec<String>>,
        metadata: Option<HashMap<String, serde_json::Value>>,
    ) {
        if let Some(s) = stability {
            self.stability = s.clamp(0.0, 1.0);
        }
        if let Some(d) = drift {
            self.drift = d.clamp(0.0, 1.0);
        }
        if let Some(c) = confidence {
            self.confidence = c.clamp(0.0, 1.0);
        }
        if let Some(r) = relationships {
            self.relationships = r;
        }
        if let Some(m) = metadata {
            self.metadata.extend(m);
        }
        self.updated_at = Utc::now();
    }
}

/// Catalog of spectral‑objects.
#[derive(Debug, Default)]
pub struct SpectralRealityModel {
    objects: HashMap<String, SpectralObject>,
}

impl SpectralRealityModel {
    /// Inserts or updates a spectral‑object.
    pub fn upsert(&mut self, obj: SpectralObject) -> &SpectralObject {
        let id = obj.id.clone();
        self.objects.insert(id, obj);
        self.objects.get(&id).unwrap()
    }

    /// Retrieves a spectral‑object by ID.
    pub fn get_by_id(&self, id: &str) -> Option<&SpectralObject> {
        self.objects.get(id)
    }

    /// Lists all objects of a given kind.
    pub fn list_by_kind(&self, kind: &SpectralKind) -> Vec<&SpectralObject> {
        self.objects
            .values()
            .filter(|o| &o.kind == kind)
            .collect()
    }

    /// Lists objects with high stability and low drift.
    pub fn list_high_stability(&self, threshold: f64) -> Vec<&SpectralObject> {
        self.objects
            .values()
            .filter(|o| o.stability >= threshold && o.drift <= 1.0 - threshold)
            .collect()
    }

    /// Lists objects from a specific domain.
    pub fn list_by_domain(&self, domain: &str) -> Vec<&SpectralObject> {
        self.objects
            .values()
            .filter(|o| o.origin.domain == domain)
            .collect()
    }

    /// Exports a snapshot of the catalog as JSON.
    pub fn snapshot(&self) -> serde_json::Value {
        serde_json::to_value(self.objects.values().collect::<Vec<&SpectralObject>>()).unwrap()
    }
}

// Example unit tests (omitted for brevity) would verify insertion, querying, and scoring logic.
